src/engraving/dom/cmd.cpp

```diff
@@ -3273,6 +3275,56 @@ void Score::cmdMirrorNoteHead()
 
 void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
 {
+    if (selection().isRange()) {
+        if (!selection().canCopy()) {
+            return;
+        }
+        ChordRest* firstCR = selection().firstChordRest();
+        if (firstCR->isGrace()) {
+            firstCR = toChordRest(firstCR->parent());
+        }
+        TDuration initialDuration = firstCR->ticks();
+        TDuration d = initialDuration.shiftRetainDots(nSteps, stepDotted);
+        if (!d.isValid()) {
+            return;
+        }
+        Fraction scale = d.ticks() / initialDuration.ticks();
+        for (ChordRest* cr : getSelectedChordRests()) {
+            Fraction newTicks = cr->ticks() * scale;
+            if (newTicks < Fraction(1, 1024)
+                || (stepDotted && cr->durationType().dots() != firstCR->durationType().dots()
+                    && !cr->isGrace())) {
+                return;
+            }
+        }
+        const muse::ByteArray mimeData(selection().mimeData());
+        XmlReader e(mimeData);
+        deleteRange(selection().startSegment(), selection().endSegment(), staff2track(selection().staffStart()),
+                    staff2track(selection().staffEnd()), selectionFilter());
+        pasteStaff(e, selection().startSegment(), selection().staffStart(), scale);
+        return;
+    }
+    if (selection().isList() && selection().elements().size() > 1) {
+        // List - act as if pressing duration toggle (distinct from range based Half/Double
+        TDuration newDuration(stepDotted
+                              ? m_is.duration().shiftRetainDots(nSteps, stepDotted)
+                              : m_is.duration().shift(nSteps));
+        m_is.duration().shiftRetainDots(nSteps, stepDotted);
+        m_is.setDuration(newDuration);
+        std::set crs = getSelectedChordRests();
+        for (auto cr : getSelectedChordRests()) {
+            changeCRlen(cr, newDuration);
+        }
+        for (auto cr : crs) {
+            EngravingItem* e = cr;
+            if (cr->isChord()) {
+                e = toChord(cr)->upNote();
+            }
+            select(e, SelectType::ADD);
+        }
+        return;
+    }
```

Comment:
Here is a fix suggestion that seemed to work quite well for me and you might want to try out. (addressing the second bug i mentioned earlier)

```suggestion
    if (selection().isList()) {
        std::set crs = getSelectedChordRests();
        for (ChordRest* cr : getSelectedChordRests()) {
            TDuration newDuration(stepDotted ? cr->durationType().fraction() * Fraction(3,(3+nSteps)) :
                                      cr->durationType().fraction() * Fraction(3-nSteps, 3+nSteps));

            changeCRlen(cr, newDuration);
        }
        for (auto cr : crs) {
            EngravingItem* e = cr;
            if (cr->isChord()) {
                e = toChord(cr)->upNote();
            }
            select(e, SelectType::ADD);
        }
        return;
    }
```

Reply:
OK, thanks, let's see...

Reply:
Well, it crashes on a ùnit test:
```
engraving_tests: ../src/engraving/dom/durationtype.cpp:414: mu::engraving::TDuration::TDuration(const mu::engraving::Fraction&, bool, int, mu::engraving::DurationType): Assertion `truncate || (fraction() - l).numerator() == 0' failed.
12/21 Test #12: engraving_tests ..................Subprocess aborted***Exception:  33.83 sec
```

Reply:
Other than that, the change seems to work as advertised, exept: W and Shift+W work only once for the list, then only one notes remains selected and is affected by further changes. Q and Shift+Q don't have this restriction.

However, that crash in the unit test is preventing this all

Reply:
Done

Reply:
the latest commit seems to combine the 2 for-loops above together which led to errors maintaining selection. I reckon the 2 for-loops shoud stay separate (like the example above) for the selecttion be correctly maintained.

Reply:
Why 2 loops? One should be enough

Reply:
I've seen the conversation in passing, and will mention that `changeCRLen()`:

![image](https://github.com/user-attachments/assets/77b3749b-54c0-4769-beaa-ea98c532bb8a)

uses a single-element selection, so it looks like after every iteration of invocation, the result will be of only one single element selection, and hence a second pass to select the finalized results make sense for this purpose. Or so it seems, of course...

Reply:
I see, so due to an in this case unwanted side effect of `changeCRLen()` we need 2 loops.

Sort of matches what I observed in https://github.com/musescore/MuseScore/pull/24385#discussion_r2024811305

Reply:
I think I got the crash reproduced, in my 3.7 version and in Debug mode:
```
Fatal: ASSERT: "truncate || (fraction() - l).numerator() == 0" in file C:\Users\Jojo\Documents\GitHub\MuseScore\libmscore\durationtype.cpp, line 437 (C:\Users\Jojo\Documents\GitHub\MuseScore\libmscore\durationtype.cpp:437, )
``

Reply:
I have found a fix for one possible crash, #10654 would have fixed that, it didn't get merged though, but got fixed differently (and better).
However, the same crash now happens again and that old PR's 2nd chunk seems to fix it (list select 2 notes of different duration and press W repeatedly).
Still not there though, the unit tests still fail...

Reply:
just in case, here is a more exact implementation for the ratios

    for (ChordRest* cr : crs) {
        Fraction upScale = stepDotted ? (cr->dots() ? Fraction(4,3) : Fraction(3,2)) : Fraction(2,1);
        Fraction downScale = stepDotted ? (cr->dots() ? Fraction(2,3) : Fraction(3,4)) : Fraction(1,2);
        Fraction scale = nSteps>0 ? downScale : upScale;
        TDuration newDuration(cr->durationType().fraction() * scale);
        changeCRlen(cr, newDuration);
    }